tests/datasets/test_streaming.py

```diff
@@ -0,0 +1,141 @@
+
+    # Test basic sample order
+    shuffle_matches = 0
+    for ix, sample in enumerate(dataset):
+        uid = sample["uid"]
+        value = sample["value"]
+        expected_uid = f"{ix:06}"
+        expected_value = 3 * ix
+        if shuffle:
+            shuffle_matches += (expected_uid == uid)
+        else:
+            assert uid == expected_uid == uid, f"sample ix={ix} has uid={uid}, expected {expected_uid}"
+            assert value == expected_value, f"sample ix={ix} has value={value}, expected {expected_value}"
+
+    # If shuffling, there should be few matches
+    # The probability of k matches in a random permutation is ~1/(e*(k!))
+    if shuffle:
+        assert shuffle_matches < 10
+
+
+@pytest.mark.parametrize(
+    "share_remote_local",
+    [
+        True,
+        pytest.param(False, marks=pytest.mark.xfail(reason="__getitem__ currently expects shards to exist")),
+    ],
+)
+def test_reader_getitem(share_remote_local: bool):
+    num_samples = 117
+    shard_size_limit = 1 << 8
+    samples, decoders = get_fake_samples_decoders(num_samples)
+    remote = write_synthetic_streaming_dataset(samples=samples, shard_size_limit=shard_size_limit)
+    if share_remote_local:
+        local = remote
+    else:
+        local = tempfile.TemporaryDirectory().name
+
+    # Build StreamingDataset
+    dataset = StreamingDataset(remote=remote, local=local, shuffle=False, decoders=decoders)
+
+    # Test retrieving random sample
+    try:
+        sample = dataset[17]
+    except Exception as e:
+        assert False, f"Unable to get random sample, got exception: {e}"
+
+
+@pytest.mark.parametrize("batch_size", [1, 2, 5, 128])
+@pytest.mark.parametrize("drop_last", [False, True])
+@pytest.mark.parametrize("num_workers", [0, 1, 8])
+@pytest.mark.parametrize("share_remote_local", [False, True])
+@pytest.mark.parametrize("shuffle", [False, True])
+def test_dataloader_single_device(batch_size: int, drop_last: bool, num_workers: int, share_remote_local: bool,
+                                  shuffle: bool):
+    num_samples = 1234
+    shard_size_limit = 1 << 8
+    samples, decoders = get_fake_samples_decoders(num_samples)
+    remote = write_synthetic_streaming_dataset(samples=samples, shard_size_limit=shard_size_limit)
+    if share_remote_local:
+        local = remote
+    else:
+        local = tempfile.TemporaryDirectory().name
+
+    # Build StreamingDataset
+    dataset = StreamingDataset(remote=remote, local=local, shuffle=shuffle, decoders=decoders, batch_size=batch_size)
+
+    # Build DataLoader
+    dataloader = DataLoader(dataset=dataset, batch_size=batch_size, num_workers=num_workers, drop_last=drop_last)
+
+    # Test dataloader length
+    expected_num_batches = (num_samples // batch_size) if drop_last else math.ceil(num_samples / batch_size)
+    num_batches = len(dataloader)
+    assert num_batches == expected_num_batches, f"Got dataloader length={num_batches} batches, expected {expected_num_batches}"
```

Comment:
Can we iterate over the data loader? That's where all the magic happens :)
Would be nice to track which samples appear and assert that each sample appears once and only once per epoch.

Reply:
Yup, the issue is torch on my Mac OSX does not support pickling of certain objects (but it works on remote Linux machines), so I got a bit stuck in the test-writing, I am planning have to work on a remote machine to finish up the DataLoader tests.